internal/caveats/run.go

```diff
@@ -98,13 +99,67 @@ func runExpression(
 	reader datastore.CaveatReader,
 	debugOption RunCaveatExpressionDebugOption,
 ) (ExpressionResult, error) {
-	if expr.GetCaveat() != nil {
-		caveat, _, err := reader.ReadCaveatByName(ctx, expr.GetCaveat().CaveatName)
-		if err != nil {
-			return nil, err
-		}
+	// Collect all referenced caveat definitions in the expression.
+	caveatNames := util.NewSet[string]()
+	collectCaveatNames(expr, caveatNames)
+
+	if caveatNames.IsEmpty() {
+		return nil, fmt.Errorf("received empty caveat expression")
+	}
+
+	// Bulk lookup all of the referenced caveat definitions.
+	caveatDefs, err := reader.LookupCaveatsWithNames(ctx, caveatNames.AsSlice())
+	if err != nil {
+		return nil, err
+	}
+
+	lc := loadedCaveats{
+		caveatDefs:          map[string]*core.CaveatDefinition{},
+		deserializedCaveats: map[string]*caveats.CompiledCaveat{},
+	}
+
+	for _, cd := range caveatDefs {
+		lc.caveatDefs[cd.Definition.GetName()] = cd.Definition
+	}
+
+	return runExpressionWithCaveats(ctx, env, expr, context, lc, debugOption)
+}
+
+type loadedCaveats struct {
+	caveatDefs          map[string]*core.CaveatDefinition
+	deserializedCaveats map[string]*caveats.CompiledCaveat
+}
+
+func (lc loadedCaveats) Get(caveatDefName string) (*core.CaveatDefinition, *caveats.CompiledCaveat, error) {
+	caveat, ok := lc.caveatDefs[caveatDefName]
+	if !ok {
+		return nil, nil, datastore.NewCaveatNameNotFoundErr(caveatDefName)
+	}
+
+	deserialized, ok := lc.deserializedCaveats[caveatDefName]
+	if ok {
+		return caveat, deserialized, nil
+	}
+
+	deserialized, err := caveats.DeserializeCaveat(caveat.SerializedExpression)
+	if err != nil {
+		return caveat, nil, err
+	}
+
+	lc.deserializedCaveats[caveatDefName] = deserialized
```

Comment:
the assumption is that this won't get modified concurrently because the caveat tree is executed serially, right?

Reply:
Correct